### DISCUSSION TITLE
how to  share location of local mathjax?

### DISCUSSION THREAD

**Discussion by berlin2123** [sphinx-doc/sphinx#12051] [Q&A]
I have muti projects,  which want to use the same local mathjax, as saying in [www.sphinx-doc.org](https://www.sphinx-doc.org/en/master/usage/extensions/math.html#module-sphinx.ext.mathjax): '_If you host more than one Sphinx documentation set on one server, it is advisable to install MathJax in a shared location._'
However, could not using a path not inside `_build/html`

such as,
```
/data/project1/_build/html           # project1
/data/project2/_build/html           # project2
/data/mathjax/es5/tex-chtml.js       # mathjax path
```
then, not work:    If set `mathjax_path = '../../../../mathjax/es5/tex-chtml.js' ` , i.e. the relative path of `/data/project1/_build/html/_static`,
it will only make browser to load matjax file from `http://IP:8000/mathjax/es5/tex-chtml.js`, instead of from   `http://IP:8000/../../../mathjax/es5/tex-chtml.js`.
However, file `/data/project1/_build/html/mathjax/es5/tex-chtml.js`  does not exist. So not work.

what will only works is copy mathjax files into a path inside `/data/project1/_build/html ` and change the `mathjax_path` too, 
However, this way does not share mathjax  bettween projects.

So, How to share?

**Comment by berlin2123** [comment]
I checked out for httpd of [using files outside of DocumentRoot](https://httpd.apache.org/docs/2.4/urlmapping.html), it could be done by `Alias` inside httpd config file. However, what I am using is sphinx-autobuild, which is based on livereload and do not support `Alias`.

But livereload supports `symbolic links`, such as:
```
ln -s /data/mathjax /data/project1/_build/html/_static/mathjax
# and in conf.py:
mathjax_path = 'mathjax/es5/tex-chtml.js'
```
will work.

BUT, This solution requires us to create `symbolic links` manually, especially after run `make clean`.
So, does it possible to automatically create `symbolic links`  in the `conf.py` file or somewhere else? just when find out the `mathjax_path` is outside of DocumentRoot.

**Comment by berlin2123** [comment]
Temporary solution for automatic copying of mathjax files:

```
mkdir /data/mathjax/_static
ln -s /data/mathjax/es5 /data/mathjax/_static/mathjax

# then, edit conf.py
html_static_path += ['../../mathjax/_static']    #  relative to configuration directory
mathjax_path = 'mathjax/tex-chtml-full.js'       #  relative to `_build/html/_static`
```
the files inside `html_static_path[i]`  will be copied into `_build/html/_static`. so It works.

Though this isn't really shared storage for mathjax, at least all projects work automatically now.

**Comment by jdillard** [comment]
I started by using this method:

```
html_static_path = ['_static', '../common/_static']

mathjax_path = 'MathJax/MathJax.js' # files located in ../common/_static
```

But, when deployed all the MathJax assets were deployed separately for each project, which slowed deploy times (lots of files) and also ate up a lot more storage over time.

Since then, I have switched to having a single project deploy with the local MathJax files like it normally would, then all other projects use `mathjax_url_schema` to reference those deployed assets (giving a little more control vs a third party CDN). It might take some custom logic to arrive at the appropriate schema if you support multiple branches/etc, but I haven't really had any issues once switching and deploy times/storage were improved (the more projects the bigger the improvement)

**Comment by berlin2123** [comment]
you are ritht, using a self-hosted `mathjax_url_schema` could work and save storge.

**Comment by berlin2123** [comment]
I'd love to know: how did the author `share mathjax` when he originally wrote that `install MathJax in a shared location` sentence.

I checked that is added by (https://github.com/sphinx-doc/sphinx/commit/1675a454ce97048870a421e87380cb38ee0f185c) into sphinx. 
I find out that is writen by Kevin Dunn.
However, I don't know how to contact him.

**Comment by berlin2123** [comment]
I'd love to know: how did the author `share mathjax` when he originally wrote that `install MathJax in a shared location` sentence.

I checked that is added by (https://github.com/sphinx-doc/sphinx/commit/1675a454ce97048870a421e87380cb38ee0f185c) into sphinx. 
I find out that is writen by Kevin Dunn.
However, I don't know how to contact  him.

**Reply by jayaddison** [reply]
I don't know whether this is what Kevin did, but: on Ubuntu Linux the JavaScript support files for `MathJax` are available in an installable system package [`libjs-mathjax`](https://packages.ubuntu.com/mantic/libjs-mathjax):

```sh
$ apt install libjs-mathjax
...
$ dpkg -L libjs-mathjax | grep MathJax.js
/usr/share/javascript/mathjax/MathJax.js
/usr/share/javascript/mathjax/unpacked/MathJax.js
```

Depending on your environment, that _might_ be a way to install the files on the webserver in a common location -- but be careful about how you provide access to those from any hosts/directories configured on the machine; you likely don't want to provide browse access to those directories, or any other directories outside the web content directories.

**Reply by berlin2123** [reply]
not mean share the mathjax in a linux file system.  
but share the installed path of mathjax bettween muti-sphinx at the same server. as said in [www.sphinx-doc.org](https://www.sphinx-doc.org/en/master/usage/extensions/math.html#module-sphinx.ext.mathjax): 'If you host more than one Sphinx documentation set on one server, it is advisable to install MathJax in a shared location.'

i.e., need the mathjax path be shared into each DocumentRoot of all sphinx projects. Otherwise it's not possible to let them both use the locally installed mathjax.

p.s.: all mathjax file have been installed fine. 
      what need is to share files, not copy them.

**Reply by jayaddison** [reply]
> i.e., need the mathjax path be shared into each DocumentRoot of all sphinx projects. Otherwise it's not possible to let them both use the locally installed mathjax.

Ok, this could help figure out the problem: what is not possible when `MathJax.js` is installed (once) in another place?

**Reply by berlin2123** [reply]
When installing mathjax into a self-built website, and other sphinx only use the js-URL of that self-built website, if the self-built website is unstable, for example, dns failure, port forwarding error, it will cause the other sphinx website can not display the math formula normally. Therefore, in order to ensure that mathjax can work independently in each sphinx site, mathjax files is needed to be placed inside DocumentRoot of each site.

**Reply by jayaddison** [reply]
@berlin2123 I agree that it must be possible to serve all of the web content (HTML, CSS, JS) required for a website to function correctly from a single webserver, to allow for the failure situations you mentioned (DNS, etc).

However: doing that doesn't require a copy of the `MathJax.js` file or a symlink to it, in every directory.  The webserver can be configured to respond to a request for `MathJax.js` on any URL by returning different content -- for example, from a Windows/Linux filesystem path.

**Reply by berlin2123** [reply]
I don't know how to configure. do you mean `Alias <SRC> <DST>` in httpd config files? 

however the webserver of `sphinx-autobuild` seems do not  support alias.

**Reply by jayaddison** [reply]
@berlin2123 it may be possible to achieve what you want without the use of aliases.

Could you check whether you are able to create a hardlink to the original/shared `MathJax.js` file on your system (not one in a Sphinx directory)?

Note: a hardlink does not _copy_ a file -- it adds another filename for it, similar to a symlink.

**Reply by berlin2123** [reply]
I can **manually** create symlink or hardlink within the system via the `ln` command (linux). Both symlink and hardlink can be used to share files from a normal user's point of view, no difference. Of course, symlinks can support folders instead of just files, and can also create links across file systems, which is more recommended.
But the problem is that I don't want to have to manually recreate the links after each `make clean` rebuild of sphinx. So I need to put the link creation process into the build process of sphinx.

**Reply by jayaddison** [reply]
>  But the problem is that I don't want to have to manually recreate the links after each `make clean` rebuild of sphinx. So I need to put the link creation process into the build process of sphinx.

Ok, this makes sense, thank you.  And to confirm: specifically it is the `MathJax.js` file that you want to use link creation during the build, yes?

**Reply by jayaddison** [reply]
(or is MathJax multiple `.js` files?)

**Reply by berlin2123** [reply]
multiple files. just from git:
```
git clone https://github.com/mathjax/MathJax.git mathjax
```
it will put all js files into `mathjax/es5`
https://docs.mathjax.org/en/latest/web/hosting.html#getting-mathjax-via-git

**Reply by jayaddison** [reply]
Ah, my question was not clear.  I meant, in an HTML page of the build dst: how many `<script src="...mathjax stuff...">` elements appear in the source of the page?

**Reply by berlin2123** [reply]
in one page, it's:
```
<script defer="defer" src="_static/mathjax/tex-chtml-full.js?v=97d8d0d7"></script>
```
seems only one js file. 

however, if I just copy or link only the `tex-chtml-full.js` file into OutDir at the same position, maths do not shows in webpages. If copy or link the `mathjax/es5` dir into OutDir, all math works fine.

**Reply by jayaddison** [reply]
> seems only one js file.

Ok, great.

> `<script defer="defer" src="_static/mathjax/tex-chtml-full.js?v=97d8d0d7"></script>`

What version of MathJax is this?  For v3.2.2 of it, I get CRC32 and SHA256:

* CRC32: `498d9c6c` (like `?v=........`)
* SHA256: `3257a831b5cba20e662dc9b840a5a88d2015aab4d50868fa251cb97f1e31a2fd`

(it may not seem relevant - but I have an idea that instead of using [capitalization](https://github.com/sphinx-doc/sphinx/blob/a2a4e3294d5a5ab853e90ebaa236ec3f526198cc/doc/usage/configuration.rst?plain=1#L1307-L1310), asking the configuration manager to include a checksum with each individual filename could be better.  only valid for files, not directories -- but we only need one file, for MathJax)

**Reply by jayaddison** [reply]
(also note: I learned this after my last comment: there is already a [`mathjax_options['integrity']`](https://www.sphinx-doc.org/en/master/usage/extensions/math.html#confval-mathjax_options) setting)

**Reply by berlin2123** [reply]
I am not submitting that PR just for mathjax. it makes more sense to link images and video files.

**Comment by jdillard** [comment]
I don't know if I would read into it to much, as implementation is likely so individual that it would hard to give more detail. Given the context, I have feeling they meant a shared hosted location, whether that's in a stand alone location or as part of a Sphinx project.

**Reply by berlin2123** [reply]
maybe.

**Comment by berlin2123** [comment]
or back to the the soft-link solution. 
if the link could run inside conf.py, all will be fine. 
But I don't know the keywords to abtain `the bulid html path` inside conf.py.

**Reply by berlin2123** [reply]
The `build path`  may not be a constant, such as running by sphinx-autobuild. However,  I can't get The `build path` while run `conf.py`, ,so manually edit conf.py my not a good solution: 
```
# edit conf.py
import os
os.system('ln -s -f path/to/mathjax  path/to/build/html/_static/mathjax')     # manually edit for build.
mathjax_path = 'mathjax/tex-chtml-full.js'
```

Here is my solve by edit the `ext/mathjax.py` file.
tested by edit the `/usr/local/lib/python3.<n>/dist-packages/sphinx/ext/mathjax.py` file, add lines from 83 line.
i.e., change 
```
    if not app.config.mathjax_path:
        msg = 'mathjax_path config value must be set for the mathjax extension to work'
        raise ExtensionError(msg)

```
to
```
    if not app.config.mathjax_path:
        msg = 'mathjax_path config value must be set for the mathjax extension to work'
        raise ExtensionError(msg)
    else:
        if ( len(app.config.mathjax_path) > 4
             and app.config.mathjax_path[0:4] != 'http' # not a Url
             and app.config.mathjax_path[0] != '/'    # not absolute path inside DocumentRoot
        ):
            import os
            mathjax_abs_path = os.path.abspath(os.path.join(app.builder.outdir, '_static', app.config.mathjax_path))
            outdir_str_len = len(os.fspath(app.builder.outdir))
            if os.fspath(app.builder.outdir)+'/' != mathjax_abs_path[0:outdir_str_len+1] :
                print('\nPath is outside, and will be linked and changed to the inside.\n')
                filename = os.path.basename(mathjax_abs_path)
                TARGET = os.path.dirname(mathjax_abs_path)
                dirname = os.path.basename(TARGET)
                LINK_NAME = os.path.abspath(os.path.join(app.builder.outdir, '_static', dirname))
                print(LINK_NAME)
                os.system('ln -s -f '+TARGET+' '+LINK_NAME)
                app.config.mathjax_path = os.path.join(dirname, filename)

```
It works fine in my test, when just setting `mathjax_path = '../../../../mathjax/_static/mathjax/tex-chtml-full.js'
` inside `conf.py`, and may need `make clean; make html`

it's just tested inside linux, may not be a good idea to push now.

**Comment by berlin2123** [comment]
mark as closed
Solutions Summary
| Solution | Advantages | Disadvantages |
| :-: | :-: |  :-: |
|  [copy files](https://github.com/orgs/sphinx-doc/discussions/12051#discussioncomment-8718953) by `html_static_path` | auto copy files into DocumentRoot; <br>just need to set the mathjax source  | may use more storage. However, its just 11M per project, my be acceptable for not too many projects |
|  [link files](https://github.com/orgs/sphinx-doc/discussions/12051#discussioncomment-8736000) by `os.system` inside  `conf.py` | auto link files into DocumentRoot; <br>link nearly do not use storage size | need to set the `path/to/mathjax` and `path/to/build/html` manually; <br>only suitable for that the build path is not changed frequently|
|  [using a self-hosted mathjax webserver](https://github.com/orgs/sphinx-doc/discussions/12051#discussioncomment-8725333)  | do not use more storage | If the mathjax webserver is unstable, it may cause math to not work properly. I.e. it is not fully guaranteed that the project can be used completely independently.|
| [edit source code](https://github.com/orgs/sphinx-doc/discussions/12051#discussioncomment-8736000) of `ext/mathjax.py` | auto link;  <br>just need to set the mathjax source  | just tested inside Linux server right now |

**Reply by berlin2123** [reply]
Other solutions are welcomed!

**Comment by berlin2123** [comment]
Solved by adding a feature to create a symlink: https://github.com/sphinx-doc/sphinx/pull/12080

After merged of this pull request, one can set conf.py as:
```
html_link_path = ['../../mathjax/_static/mathjax']    #  relative to configuration directory
mathjax_path = '_DoNotEditHerein/mathjax/tex-chtml-full.js'
```
or
```
html_link_path = [ ['../../mathjax/es5', '_static/_DoNotEditHerein/mathjax']  ]
mathjax_path = '_DoNotEditHerein/mathjax/tex-chtml-full.js'
```
